### PR TITLE
Implement BIP39 (Mnemonic code for generating deterministic keys)

### PR DESCRIPTION
**Pull request by levonpetrosyan93**
Fixes https://github.com/zcoinofficial/zcoin/issues/732
Implementation of #559 

https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki

High level summary:
- 12 and 24 words supported
- passphrase supported but optional.
- By default, new wallets will be created with mnemonic. Users can choose to disable a mnemonic wallet with `usemnemonic=0`.
- In `zcoind` a 24 word mnemonic with no passphrase is created by default. This is also the default setting at the Qt screen
- Qt users are shown the mnemonic page on startup, if:
    - new wallet being created
    - `usemnemonic=1` (default. meaning the page is not shown if `usemnemonic=0` passed)
- `importwallet` and `importprivkey` block with a mnemonic enabled wallet. A suggestion is given to create a non-mnemonic wallet if import is need. The purpose of this is to have a single mnemonic backup that is not complicated by other imports. (The keys would still import into the wallet if this check was removed, but presumably whoever was to do this would be aware of the repercussions).

- conf setting interactions are handled

- This PR also removes the code that creates a new Master key following encryption, either for a mnemonic wallet or older BIP32/44 wallets. This feature was introduced by us when Zcoin launched (ie. it was not a feature of Bitcoin 0.13 or any other version ); it provides no extra security benefit and only complicates the wallet code. Warning messages have been updated to reflect this change.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1377,6 +1387,36 @@ CPubKey CWallet::GenerateNewHDMasterKey() {
...
+    else {
+        LogPrintf("CWallet::GenerateNewHDChain -- Generating new HD Chain\n");
+
+        std::string mnemonic = GetArg("-mnemonic", "");
+        std::string mnemonicPassphrase = GetArg("-mnemonicpassphrase", "");
+
+        SecureString secureMnemonic(mnemonic.begin(), mnemonic.end());
+        SecureString securePassphrase(mnemonicPassphrase.begin(), mnemonicPassphrase.end());
+
+        if (!newHdChain.SetMnemonic(secureMnemonic, securePassphrase, true))
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: Yes It is called, and in than case new mnemonic is being generated.

File: src/wallet/hdchain.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+{
+    SecureString mnemonicNew = mnemonic_;
+    if(setMasterKeyID) {
+        if(mnemonicNew.empty())
+            mnemonicNew = Mnemonic::mnemonic_generate(256);
+        if(!Mnemonic::mnemonic_check(mnemonicNew))
+            throw std::runtime_error(std::string(__func__) + ": mnemonic is invalid");
+        Mnemonic::mnemonic_to_seed(mnemonicNew, passPhrase_, seed);
+
+        masterKeyID = CKeyID(Hash160(seed.begin(), seed.end()));
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/wallet/hdchain.cpp]: Did a research on this, I think we need to somehow use 256bit for mnemonic and keep 160 for old versions, but need to so more investigation how make this compatible with HW wallets.

File: src/wallet/wallet.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: Now when -mnemonic is not set, that means user did not provided memonic words for recovery, than new mnemonic is not generated by default. I mean in that case new wallet es being created. I we want to allow choose between legacy seed creation or mnemonic, than we can add another argument for that.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1377,6 +1387,36 @@ CPubKey CWallet::GenerateNewHDMasterKey() {
     return pubkey;
 }
 
+void CWallet::GenerateNewHDChain(){
+    CHDChain newHdChain;
+
+    std::string strSeed = GetArg("-hdseed", "not hex");
```
[the remark this answers is not in the document]
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: I looked several other places, they did same way. I think better to have this also to allow users recover hd wallets crated before mnemonic version.

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -702,8 +702,36 @@ UniValue dumpwallet(const UniValue& params, bool fHelp)
     file << "\n";
 
     // add the base58check encoded extended master if the wallet uses HD
-    CKeyID masterKeyID = pwalletMain->GetHDChain().masterKeyID;
-    if (!masterKeyID.IsNull())
+    CHDChain chain = pwalletMain->GetHDChain();
+    CKeyID masterKeyID = chain.masterKeyID;
+    if(!chain.IsNull())
```
Comment (riordant) [inline src/wallet/rpcdump.cpp]: @levonpetrosyan93 I think there should also be a condition check here for HDChain version >= `VERSION_WITH_BIP39`

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -698,11 +702,12 @@ bool CWallet::EncryptWallet(const SecureString &strWalletPassphrase) {
             pwalletdbEncryption = NULL;
         }
 
-        Lock();
-        Unlock(strWalletPassphrase, true);
-
         // if we are using HD, replace the HD master key (seed) with a new one
-        if (!hdChain.masterKeyID.IsNull()) {
+        if(!hdChain.IsNull()) {
```
Comment (riordant) [inline src/wallet/wallet.cpp]: Same here: `hdChain.nVersion >= CHDChain::VERSION_WITH_BIP39`

File: src/wallet/wallet.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: I agree with the current implementation, and don't think we should use legacy wallets with mnemonic wallet type anymore

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -4037,6 +4037,7 @@ extern UniValue importprivkey(const UniValue& params, bool fHelp);
 extern UniValue importaddress(const UniValue& params, bool fHelp);
 extern UniValue importpubkey(const UniValue& params, bool fHelp);
 extern UniValue dumpwallet_zcoin(const UniValue& params, bool fHelp);
+extern UniValue dumpmnemonic(const UniValue& params, bool fHelp);
```
Comment (riordant) [inline src/wallet/rpcwallet.cpp]: I also think an `importmnemonic` function is needed. Make sure to check for wallet version

Comment (levonpetrosyan93) [reply src/wallet/rpcwallet.cpp]: I think better to allow  inputting mnemonic only during first startup.

File: src/wallet/rpcdump.cpp (hunk 4 above)
Comment (levonpetrosyan93) [reply src/wallet/rpcdump.cpp]: Good catch. Thank you.

File: src/wallet/hdchain.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/hdchain.cpp]: @psolstice I believe this can be marked resolved?

File: src/wallet/wallet.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: @psolstice Also this? The idea is that new wallets will always generate a mnemonic

**Comment by qashban4** [comment]
@riordant @ultimaweapon can you review so we can merge.

**Comment by ultimaweapon** [comment]
Nope I can't. Don't have time to review this due to need to focus on Exodus.

**Comment by qashban4** [comment]
Review to be complete by 2019-09-26 for further merge.

**Comment by qashban4** [comment]
@levonpetrosyan93 to test it with HW wallet during W40
@psolstice to review and test on HW wallet during W40

**Comment by qashban4** [comment]
@psolstice update your review and test result on HW wallet

**Comment by qashban4** [comment]
@levonpetrosyan93 to finalize UI 
Once levon tests it with hw key and tells test procedure 
@psolstice needs less than a day to test and finalize review.
Aim to close it during middle of W42.

**Comment by qashban4** [comment]
@psolstice lets finalize this week (W42) so we can merge it to master.

**Comment by qashban4** [comment]
Instructions for testing mnemonic with hardware ui
1. Create a new wallet
2. Open zcoin_qt -> Dubug window -> Console
3. type "dumpmnemonic" command
4. copy mnemonic words
5. Open https://wallet.trezor.io
6. Wipe device if needed and reconnect it
7. Click on "Recover Wallet"
8. Input words (it will ask by random order)
9. Once wallet is recovered, create addresses and compare to  addresses in zcoin_qt

**Comment by qashban4** [comment]
@riordant @psolstice lets review and conclude to merge to master Thanks.

File: src/qt/notifymnemonic.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+void NotifyMnemonic::notify()
+{
+    NotifyMnemonic notify;
+    SecureString mnemonic;
+    SecureString passphrase;
+    pwalletMain->GetHDChain().GetMnemonic(mnemonic, passphrase);
+    notify.ui->mnemonic->setText(mnemonic.c_str());
+    if(notify.exec())
```
Comment (riordant) [inline src/qt/notifymnemonic.cpp]: This call is causing a failure (on Mac at least). 
When creating a new wallet, this function is called. It looks like the `exec` function doesn't exist.
Output:
```
************************
UNKNOWN EXCEPTION       
zcoin in Runaway exception  
```

GUI says this:
`A fatal error occurred. Zcoin can no longer continue safely and will quit.`

File: src/wallet/hdchain.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
+#ifndef ZCOIN_HDCHAIN_H
+#define ZCOIN_HDCHAIN_H
+
+#include "key.h"
+
+/* simple HD chain data model */
+class CHDChain
```
Comment (psolstice) [inline src/wallet/hdchain.h]: You absolutely cannot move CHDChain to another file because it's bitcoin code. It will create a lot of problems when upgrading to newer bitcoin cores

File: src/wallet/hdchain.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            READWRITE(nExternalChainCounters);
+            nExternalChainCounters.resize(N_CHANGES);
+        }
+        if(this->nVersion >= VERSION_WITH_BIP39) {
+            READWRITE(mnemonic);
+            READWRITE(passPhrase);
+            READWRITE(seed);
+            READWRITE(fIsCrypted);
+            READWRITE(f12Words);
+        }
```
Comment (psolstice) [inline src/wallet/hdchain.h]: Why do we save/restore mnemonic, passPhrase, seed etc? Do we really really store password to the wallet in plain text next to it?

File: src/wallet/hdchain.h (hunk 8 above)
Comment (levonpetrosyan93) [reply src/wallet/hdchain.h]: In Dash it is done the same way.

File: src/wallet/hdchain.h (hunk 9 above)
Comment (levonpetrosyan93) [reply src/wallet/hdchain.h]: we save it to be able to allow user to dump it any time.

Comment (psolstice) [reply src/wallet/hdchain.h]: no! we might print it *once* during new wallet initialization. Sorry to overlook it earlier but there shouldn't be postfactum dumpmnemonic command at all. Look how it's done when setting up Trezor one for example.
The way you do it effectively ruins every wallet security. Only flags/counters/key ids can be saved

File: src/wallet/hdchain.h (hunk 8 above)
Comment (psolstice) [reply src/wallet/hdchain.h]: not a good example because they implemented HD stuff to 0.12 version of their bitcoin core, we're already at 0.13 and should not mess with that

File: src/wallet/hdchain.h (hunk 9 above)
Comment (psolstice) [reply src/wallet/hdchain.h]: I saw that's how dash is doing it but that's definitely not how it should be done

File: src/wallet/hdchain.h (hunk 8 above)
Comment (levonpetrosyan93) [reply src/wallet/hdchain.h]: ok, good, will move it back.

File: src/wallet/hdchain.h (hunk 9 above)
Comment (psolstice) [reply src/wallet/hdchain.h]: Got to the bottom of it. Dash get its own implementation of `CHDChain` and it's encrypted/decrypted when saved into the `wallet.db`. Your implementation is a mix between that of bitcoin and dash. And that means all of these fields are saved to the wallet.db **unencrypted**. Which means a huge security hole

Comment (levonpetrosyan93) [reply src/wallet/hdchain.h]: It is being encrypted when you encrypt the wallet, it is being encrypted in our case also.

File: src/qt/notifymnemonic.cpp (hunk 7 above)
Comment (levonpetrosyan93) [reply src/qt/notifymnemonic.cpp]: fixed

File: src/wallet/hdchain.h (hunk 9 above)
Comment (psolstice) [reply src/wallet/hdchain.h]: nope, only keys are encrypted, everything you write to the wallet with `Write` is not

File: src/wallet/mnemoniccontainer.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    bool f12Words;
+
+    MnemonicConatiner() { SetNull(); }
+
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion)
+    {
+            READWRITE(mnemonic);
+            READWRITE(seed);
```
Comment (psolstice) [inline src/wallet/mnemoniccontainer.h]: Why do we need to save seed? The only purpose of `MnemonicContainer` is to save mnemonic for later recovery by user. Seed is saved within `wallet.dat` by bitcoin code, no need to duplicate it here.

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -1462,6 +1522,89 @@ bool CWallet::SetHDChain(const CHDChain &chain, bool memonly) {
...
+        if (!mnemonicConatiner.SetMnemonic(secureCryptedMnemonic, secureCryptedPassphrase, false))
+            return false;
+    }
+
+    mnemonicConatiner.SetCrypted(true);
+
+    return true;
+}
+
+bool CWallet::DecryptMnemonicConatiner(MnemonicConatiner& mnConatiner) const
```
Comment (psolstice) [inline src/wallet/wallet.cpp]: Typo. You need to replace Conatiner with Container all over the code

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -677,10 +689,10 @@ bool CWallet::EncryptWallet(const SecureString &strWalletPassphrase) {
     if (IsCrypted())
         return false;
 
-    CKeyingMaterial vMasterKey;
```
Comment (psolstice) [inline src/wallet/wallet.cpp]: I am strongly against renaming like that. It makes any core upgrade a lot harder. If there is a conflict with another variable introduced in this PR that variable needs renaming, not the original one

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -1462,6 +1522,89 @@ bool CWallet::SetHDChain(const CHDChain &chain, bool memonly) {
...
+    if (mnemonicConatiner.GetMnemonic(cryptedMnemonic)) {
+        SecureVector vectorMnemonic;
+
+        std::vector<unsigned char> CryptedMnemonic(cryptedMnemonic.begin(), cryptedMnemonic.end());
+        if (!CryptedMnemonic.empty() && !DecryptSecret(vMasterKey, CryptedMnemonic, id, vectorMnemonic))
+            return false;
+
+        SecureString mnemonic(vectorMnemonic.begin(), vectorMnemonic.end());
+        SecureString passphrase;
+        if (!mnConatiner.SetMnemonic(mnemonic, passphrase, false))
```
Comment (psolstice) [inline src/wallet/wallet.cpp]: Passing empty passphrase isn't looking good or obvious. There is no need to derive seed ad this point (the only goal of using `DecryptMnemonicContainer` is to display mnemonic to the user, password is something only the user knows about). Or at least do it without adding new local variable

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -1462,6 +1522,89 @@ bool CWallet::SetHDChain(const CHDChain &chain, bool memonly) {
...
+            return false;
+    }
+
+    mnConatiner.SetCrypted(false);
```
Comment (psolstice) [inline src/wallet/wallet.cpp]: Does it remain unencrypted in memory until client quits? Can we use the same time-based mechanism to remove it from memory when the wallet locks on timeout? Or after there is no need in reading mnemonic anymore?

Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: It is being deleleted, when going out of scope.

File: src/wallet/wallet.cpp (hunk 13 above)
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: when we pass false it only set's mnemonic, but may be better to create seperate function for that. Will do tat way.

File: src/wallet/wallet.cpp (hunk 11 above)
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: thank you, will replace.

File: src/wallet/wallet.cpp (hunk 12 above)
Comment (levonpetrosyan93) [reply src/wallet/wallet.cpp]: ok will try to find a better solution.

File: src/wallet/mnemoniccontainer.h (hunk 10 above)
Comment (levonpetrosyan93) [reply src/wallet/mnemoniccontainer.h]: It only being kept here.

Comment (psolstice) [reply src/wallet/mnemoniccontainer.h]: Keeping secret data when unnecessary is generally not a good idea. And seed is essentially ultimate attacker's goal

Comment (levonpetrosyan93) [reply src/wallet/mnemoniccontainer.h]: I got your point but we are keeping seed here for deriving keys, please take a look on GenerateNewKey() and GetKeyFromKeypath() functions.

File: src/wallet/crypter.cpp [hunk 15]
```diff
@@ -113,7 +113,7 @@ bool CCrypter::Decrypt(const std::vector<unsigned char>& vchCiphertext, CKeyingM
 }
 
 
-static bool EncryptSecret(const CKeyingMaterial& vMasterKey, const CKeyingMaterial &vchPlaintext, const uint256& nIV, std::vector<unsigned char> &vchCiphertext)
+bool CCryptoKeyStore::EncryptSecret(const CKeyingMaterial& vMasterKey, const CKeyingMaterial &vchPlaintext, const uint256& nIV, std::vector<unsigned char> &vchCiphertext)
```
Comment (psolstice) [inline src/wallet/crypter.cpp]: Sorry for not spotting this earlier. Modifying bitcoin code like that is not something we'd want to do. On the other hand adding another method to the `CCryptoKeyStore` is ok. This way you can even revert the change that makes `vMasterKey` protected instead of private

File: src/wallet/mnemoniccontainer.h (hunk 10 above)
Comment (psolstice) [reply src/wallet/mnemoniccontainer.h]: I still don't get it. When initializing wallet with mnemonic we generate new seed and from that moment it is in the hands of bitcoin code. Why do we need to do our own way of generating new key? Why can't bitcoin code generate one for us?

File: src/wallet/crypter.cpp (hunk 15 above)
Comment (levonpetrosyan93) [reply src/wallet/crypter.cpp]: Ok, will do that way. Thank you.

File: src/wallet/mnemoniccontainer.h (hunk 10 above)
Comment (levonpetrosyan93) [reply src/wallet/mnemoniccontainer.h]: Mnemonic seed is being generated 64 byte, but Key we can set only 32, and using other hacks is generating addresses not compatible to hardware wallets.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -752,7 +764,11 @@ bool CWallet::EncryptWallet(const SecureString &strWalletPassphrase) {
         Unlock(strWalletPassphrase, true);
 
         // if we are using HD, replace the HD master key (seed) with a new one
-        if (!hdChain.masterKeyID.IsNull()) {
+        if(!mnemonicContainer.IsNull() && hdChain.nVersion >= CHDChain::VERSION_WITH_BIP39) {
+            assert(EncryptMnemonicContainer(vMasterKey));
+            SetMinVersion(FEATURE_HD);
+            assert(SetMnemonicContainer(mnemonicContainer, false));
+        } else if (!hdChain.masterKeyID.IsNull()) {
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: Resolved following merge of my branch.

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -1427,6 +1449,44 @@ CPubKey CWallet::GenerateNewHDMasterKey() {
...
+    bool isHDSeedSet = strSeed != "not hex";
+
+    if(isHDSeedSet && IsHex(strSeed)) {
+        std::vector<unsigned char> seed = ParseHex(strSeed);
+        if (!mnContainer.SetSeed(SecureVector(seed.begin(), seed.end())))
+            throw std::runtime_error(std::string(__func__) + ": SetSeed failed");
+        newHdChain.masterKeyID = CKeyID(Hash160(seed.begin(), seed.end()));
+    }
+    else {
+        LogPrintf("CWallet::GenerateNewHDChain -- Generating new MnemonicContainer\n");
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: Resolved following merge of my branch.
